enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -0,0 +1,306 @@
+
+#### Install
+
+A comparison of installing the implementation of this proposal via OLM vs. CVO was done.  The net result was 
+to go with CVO.
+
+Some specifics:
+- During the review for the ["Installation of CSI drivers that replace in-tree drivers](https://github.com/openshift/enhancements/blob/master/enhancements/storage/csi-driver-install.md)
+a question was posed about whether the exact type of CSI driver this proposal is describing would use the method 
+described in that document (namely OLM).  The recommendation from the [storage SMEs was to go with CVO](https://github.com/openshift/enhancements/pull/139/files#r391761362)
+- In particular, using CVO and having it be a day 1 install operation meets the ease of use targets we are trying to
+reach with this proposal
+- Crafting dependencies between this proposal and [related proposals](https://github.com/openshift/enhancements/pull/214)
+at the time of this writing is immature.
+
+#### Brief overview of CSI and the main K8S hook points
+
+ - a [CSI plugin](https://github.com/container-storage-interface/spec) is a gRPC endpoint that implements CSI services and
+ allows storage vendors to develop a storage solution that works across a number of container orchestration systems.
+ - though for our purposes, we will not provide a generic CSI plugin, but rather a special case one that is only intended
+ to work on OpenShift
+ - also for our purposes, `plugin`, `driver`, and `endpoint` seem to be the same thing based on which document you are
+ reading.
+ - k8s provides Kubernetes CSI Sidecar Containers: a set of standard containers that aim to simplify the development and 
+ deployment of CSI Drivers on Kubernetes.
+ - There are CSI volume and persistent volume types in k8s.
+ - per [k8s volume docs](https://kubernetes.io/docs/concepts/storage/volumes/#csi) "the CSI types do not support direct
+ reference from a Pod and may only be referenced in a Pod via a `PersistentVolumeClaim` object" (direct quote)
+ - However, @gnufied tells us that documentation is out of date and incorrect.  To quote him:  That is where [inline 
+ CSI volumes come](https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html)
+ - To that end, one of the sidecar containers, the [external provisioner](https://kubernetes-csi.github.io/docs/external-provisioner.html),
+ helps with this
+ - It watches Kubernetes `PersistentVolumeClaim` objects, and if the PVC 
+ references a Kubernetes StorageClass, and the name in the provisioner field of the storage class matches the name 
+ returned by the specified CSI endpoint `GetPluginInfo` call, it calls the CSI endpoint `CreateVolume` method to provision
+ the new volume.
+ - there are also lifecycle methods for destroy, resize, etc.
+ 
+ #### Some high level thoughts on what we need to inject/consume content on every host/node in an OpenShift cluster
+   
+ - the CSI plugin would be hosted as a `DaemonSet` on the k8s cluster, to get per node/host granularity
```

Comment:
On security note: what Secrets can this DaemonSet access? It probably needs access to entitlements and similar cross-namespace Secrets, but it should not get access to all Secret in the cluster.

Reply:
it will have access to all secrets in the cluster because the admin will be able to use it to inject any secret from any namespace into any pod in any namespace, not just entitlements.

The security will be provided via admission logic which constrains who can create/mount these volumes in their pods (outside the scope of this proposal which is just about the CSI driver itself)

Reply:
So from my perspective, when consuming comments on this and #214, reviewers' perspectives on when to reveal particulars from #214 here and when to reveal particulars from this proposal there has been a bit of a moving target.

With that as a background @bparees , should I try to give some sort of summary of the admission logic piece here to help address readers' concerns like what @jsafrane listed here?

Reply:
Or just say "see ... for how security will be handled" ?

Reply:
> Or just say "see ... for how security will be handled" ?

this one.  I would just explicitly state that this CSI driver is not responsible for security, it enables god mode and admins will need to deploy it in conjunction with something that limits who can use it/how it can be used.

Reply:
that works for me @bparees thanks